### ISSUE TITLE
Passing Changes to be commited to underline testrunner for new strategy

### ISSUE THREAD

**Issue by dipak-pawar** [arquillian/smart-testing#24]
*Steps to reproduce*
- Take [functional-tests/changed-ftest](https://github.com/arquillian/smart-testing/tree/41ade53a907ea284e8da3dee11b909be4aac6fd3/functional-tests/changed-ftest) project.
- Change already committed any test i.e. FirstTest. Add it to staging area using git add file_name
- Run the build using `mvn test`

You will see changes to be committed are not running. It's excluding as it's in staged area.

If you try to add new file which is not tracked(adding for the first time) & added by using `git add` in staging area is not excluding.

**Comment by bartoszmajsak** [comment]
What is the expected behaviour for you?

**Comment by dipak-pawar** [comment]
Ideally this file should not be skipped.

**Comment by dipak-pawar** [comment]
Or we should keep track of it either in `new` or `changed`. Currently I am looking it only for `new` strategy.

**Comment by bartoszmajsak** [comment]
Another a bit misleading fact is that module you are referring to has following strategies enables:

```
<strategies>new,changed</strategies>
```

Maybe we should break them apart.

**Comment by dipak-pawar** [comment]
@bartoszmajsak no, I have tested it by breaking them & using it only `new`. It's not working for staged file which is already tracked in previous commits.

**Comment by dipak-pawar** [comment]
@bartoszmajsak , @MatousJobanek @lordofthejars Wdyt about this?

**Comment by MatousJobanek** [comment]
Hi, 
from my point of view, if no commit range is specified then the local changes should be run (when you have `changed` strategy. So, as you have described @dipak-pawar the changed file should not be skipped and should be included in the set of tests. `new` I understand as newly created files

**Comment by lordofthejars** [comment]
Just one thing, local changes (that is local files and files that are in git index area) are only detected with `new` strategy.

**Comment by bartoszmajsak** [comment]
> Just one thing, local changes (that is local files and files that are in git index area) are only detected with new strategy.

For me they are `new` so they are detected by the right thing

**Comment by lordofthejars** [comment]
yes, just commenting this fact.

**Comment by dipak-pawar** [comment]
**Strategy** `new`

Run tests using `mvn clean test -Dgit.previous.commit=$PREV_COMMIT -Dgit.commit=$COMMIT`

If you have two changes in given commit ranges -> added new file(FirstTest), modified file(SecondTest).

If you try to run tests with commit range then it will run all new tests added in commit ranges. In our case it will run only 'FirstTest'

**Strategy** `changed`

Run tests using `mvn clean test -Dgit.previous.commit=$PREV_COMMIT -Dgit.commit=$COMMIT`

If you have two changes in given commit ranges -> added new file(FirstTest), modified file(SecondTest).

If you try to run tests with commit range, then it will run all modified/changed tests added in commit ranges. In our case it will run only 'SecondTest'

Everything above looks fine.

**local changes**

Now think for following local changes:
You have new file as untracked(ThirdTest), any modified file which is not staged(SecondTest), any modified file which is staged(FirstTest). https://gist.github.com/dipak-pawar/a9bf3ba5b987956429730a9342fa261b

**Strategy**: `new` 
Run tests using `mvn clean test`
Current behviour - it's detecting all local changes i.e. untracked changes(ThirdTest), modified changes - not staged for commit(SecondTest), new staged chnages(ForthTest) except modified staged changes(FirstTest.java).

Here we are running SecondTest,ThirdTest, ForthTest & ignoring FirstTest.

**Strategy**: `changed`, run tests using `mvn clean test`
Current behviour - No tests to run

So in case of local changes with strategy `new` we should track only new changes from untracked(ThirdTest) & staged area(ForthTest), 
in case of strategy `changed` we should track all modified changes from tracked(SecondTest) & staged area(FirstTest)

OR

make new strategy for local changes something like `local-new` + `local-changed` = `local`

WDYT?

**Comment by bartoszmajsak** [comment]
I think we should keep it simple and have following behaviour (with an example based on your scenario and pretty much close to what you are proposing):

### New

Tracks new tests:

  * `new file` in the commit 
  * `new file` in the staged area
  * untracked files

In this scenario both `ForthTest` and `ThirdTest` should be in.

### Changed

Tracks modification only of changed files (so those which were existing in the repository before we started doing our local changes).

  * `modified` in the commit 
  * `modified` in the staged area

In this scenario both `FirstTest` and `SecondTest` should be in.

# Dealing with commit ranges

When it comes to commit ranges we can think about three scenarios:

  * no range specified
    * we only look at `HEAD`
  * from 
    * we look at `from...HEAD`
  * full range (typical CI scenario)


Are we good with it?

**Comment by dipak-pawar** [comment]
We are good with it. Thanks @bartoszmajsak
